packages/nextjs/pages/blockexplorer.tsx

```diff
@@ -0,0 +1,172 @@
+import type { NextPage } from "next";
+import { ArrowLeftIcon, ArrowRightIcon } from "@heroicons/react/24/outline";
+import { SearchBar } from "~~/components/blockexplorer/SearchBar";
+import { TransactionsTable } from "~~/components/blockexplorer/TransactionsTable";
+import deployedContracts from "~~/generated/deployedContracts";
+import { getTargetNetwork } from "~~/utils/scaffold-eth";
+
+// TODO: Add OPCODES and maybe STORAGE for contracts
+// TODO: Better pagination(some pages have less than BLOCKS_PER_PAGE blocks)
+// TODO: Refactor, seperate the table and the search bar to a component
+// TODO: Don't do everything in a loop... Please...
+// TODO: try useMemo to avoid unnecessary re-renders
+// TODO: Error handling for API calls
+// TODO: Add tests
+
+type TransactionWithFunction = TransactionResponse & { functionName?: string };
+
+const BLOCKS_PER_PAGE = 20;
+
+const Blockexplorer: NextPage = () => {
+  const [blocks, setBlocks] = useState<BlockWithTransactions[]>([]);
+  const [transactionReceipts, setTransactionReceipts] = useState<{
+    [key: string]: ethers.providers.TransactionReceipt;
+  }>({});
+  const [currentPage, setCurrentPage] = useState(0);
+  const [totalBlocks, setTotalBlocks] = useState(0);
+  const [searchInput, setSearchInput] = useState("");
+
+  const configuredNetwork = getTargetNetwork();
+  const router = useRouter();
+
+  const provider = new ethers.providers.JsonRpcProvider("http://localhost:8545");
+
+  const fetchBlocks = async () => {
+    const blockNumber = await provider.getBlockNumber();
+    setTotalBlocks(blockNumber);
+    const blocks: BlockWithTransactions[] = [];
+    const receipts: { [key: string]: ethers.providers.TransactionReceipt } = {};
+    const startingBlock = blockNumber - currentPage * BLOCKS_PER_PAGE;
+
+    for (let i = 0; i < BLOCKS_PER_PAGE; i++) {
+      const blockNumberToFetch = startingBlock - i;
+      if (blockNumberToFetch < 0) {
+        break;
+      }
+
+      const block = (await provider.getBlockWithTransactions(blockNumberToFetch)) as BlockWithTransactions & {
+        transactions: TransactionWithFunction[];
+      };
+
+      blocks.push(block);
+
+      const chain = deployedContracts[31337][0];
+      const interfaces: { [contractName: string]: ethers.utils.Interface } = {};
+
+      for (const [contractName, contract] of Object.entries(chain.contracts)) {
+        interfaces[contractName] = new ethers.utils.Interface(contract.abi);
+      }
+
+      for (const tx of block.transactions as TransactionWithFunction[]) {
+        const receipt = await provider.getTransactionReceipt(tx.hash);
+        receipts[tx.hash] = receipt;
+
+        if (tx.data.length >= 10) {
+          for (const [contractName, contractInterface] of Object.entries(interfaces)) {
+            try {
+              const decodedData = contractInterface.parseTransaction({ data: tx.data });
+              tx.functionName = `${contractName}: ${decodedData.name}`;
+              break;
```

Comment:
Found an edge case with this logic when functions have same signature in different contract it shows the first contractName which matches : 

https://github.com/scaffold-eth/scaffold-eth-2/assets/80153681/c093e54f-36ee-47d7-8fa4-a801f8c9e73e